genet/schedule_elements.py

```diff
@@ -1210,6 +1212,74 @@ def route_trips_to_dataframe(self, gtfs_day='19700101'):
         return df
 
+    def unused_vehicles(self):
+        """
+        A scenario change to the network may result in changes to vehicle assignments, with some vehicles not
+        being used anymore. This method checks if any of the vehicles are missing (i.e. exist in Schedule.vehicles,
+        but are not used by services) and returns a list of these vehicles' IDs, if there are any.
+        It also logs a warning which says whether any unused vehicles have been found.
+        """
+        # :param vehicles = dictionary of vehicle IDs from Route objects
+        # Looks like this: {veh_id : {'type': 'bus'}}
+        # e.g.  {'fun_bus_1': {'type': 'bus'}, 'fun_bus_2': {'type': 'bus'}, 'some_bus_2': {'type': 'bus'}}
+
+        existing_vehicles = set(self.vehicles.keys())
+        used_vehicles = self.route_trips_to_dataframe()
+        used_vehicles = set(used_vehicles['vehicle_id'])
+
+        unused_vehicles = existing_vehicles - used_vehicles
+
+        if len(unused_vehicles) == 0:
+            logging.info('All vehicles are being used.')
+        else:
+            logging.warning(str(len(unused_vehicles)) + ' unused vehicles have been found.')
+
+        return unused_vehicles
+
+    def check_vehicle_uniqueness(self):
+        """
+        In MATSim, trips can share vehicles, but his may or may not be intended, e.g. it could result from a
+        scenario change and be undesirable, leading to simulation not working correctly.
+        This method checks if a vehicle ID is being used by two or more different trips, and then returns
+        a dictionary of vehicle IDs together with trips for which they are being used.
+        It also logs a warning which says whether any vehicles are being used for multiple trips.
+        """
+        trips_df = self.route_trips_to_dataframe()
+        trips_df = trips_df[['trip_id', 'vehicle_id']]
+
+        trips_dict = {}
+        for i in range(0, len(trips_df)):
+            key = trips_df['trip_id'].iloc[i]
+            value = trips_df['vehicle_id'].iloc[i]
+            trips_dict[key] = value
+
+        # finding duplicate values from dictionary by flipping keys and values
+        flipped = {}
+        not_unique_list = []
+
+        for k, v in trips_dict.items():
+            if v not in flipped:
+                flipped[v] = [k]
+            else:
+                not_unique_list.append(v)
+                flipped[v].append(k)
+
+        duplicates_dict = {}
+        for key in not_unique_list:
+            values = flipped[key]
+            duplicates_dict[key] = []
+            for val in values:
+                duplicates_dict[key].append(val)
+
+        if len(duplicates_dict) == 0:
+            logging.info('No vehicles being used for multiple trips have been found.')
+        else:
+            non_unique_vehicles = duplicates_dict.keys()
+            printing_list = str(non_unique_vehicles)[11:-2]
```

Comment:
Because when you tell a logging.warning() function (line 1279) to print a list, it prints it with brackets and name, so I had to get rid of these to just print the actual contents of the list